iree/turbine/kernel/wave/cache.py

```diff
@@ -0,0 +1,233 @@
+import os
+import hashlib
+import inspect
+import json
+import torch
+
+from collections import OrderedDict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Callable
+
+from .constraints import Constraint
+from ..compiler.kernel_codegen import KernelBufferUsage
+from .._support.indexing import IndexExpr
+from .utils import invoke_vmfb, _read_file, _write_file
+
+default_cache_base_dir = f"{str(Path.home())}/.wave"
+CACHE_BASE_DIR = str(os.environ.get("WAVE_CACHE_DIR", default_cache_base_dir))
+WAVE_ALWAYS_COMPILE = int(os.environ.get("WAVE_ALWAYS_COMPILE", 0))
+WAVE_CACHE_LIMIT = int(os.environ.get("WAVE_CACHE_LIMIT", 16))
+
+
+@dataclass
+class WaveCache:
+    """
+    Dataclass/Struct that stores necessary information S.T we can
+    reconstruct and call the "cached" kernel.
+    """
+
+    cache_id: str
+    kernel_sig: tuple[KernelBufferUsage]
+    vmfb: bytes
+
+    @property
+    def module_op(self):
+        filepath = f"{CACHE_BASE_DIR}/{self.cache_id}/{self.cache_id}.mlir"
+        if not os.path.exists(filepath):
+            raise ValueError("Failed to find module op MLIR for cached kernel.")
+        with open(filepath, "r") as f:
+            module_str = f.read()
+        return module_str
+
+
+class WaveCacheManager(object):
+    """
+    Wave cache manager has two main components/cache:
+
+    1. Session/Online cache - This is the main cache that our compiler and runtime will load from and store to. It is
+    essentially a dict that uses the kernel hash as keys and the WaveCache as values. We added LRU functionality with limits
+    for number of kernel cached here, because this lives on RAM, and we wouldn't want to run OOM.
+
+    2. File/Offline cache - This cache is essential for loading saved/compiled cache between sessions/runs. This is done
+    by storing vital kernel information(vmfb, kernel_sig, and mlir) to CACHE_BASE_DIR/kernel_hash directory. If said kernel
+    is queried during a new run and does not exist on session/online cache yet, we'd load files from the kernel_hash directory
+    and reconstruct the WaveCache from it.
+    """
+
+    def __init__(self):
+        self.file_cache: set[str] = set()
+        self.session_cache: OrderedDict[str, WaveCache] = OrderedDict()
+        self.update_file_cache()
+
+    def get_hash(
+        self,
+        constraints: list[Constraint],
+        kernel_body: Callable,
+        hyperparams: dict[IndexExpr, Any],
+        dynamic_symbols: list[IndexExpr, Any],
+        config: dict[str, str],
+        use_scheduling: bool,
+        use_scheduling_barriers: bool,
+        run_bench: bool,
+    ):
+        """
+        Get a unique identifier for a given kernel.
+        """
+        key = [
+            inspect.getsource(kernel_body),
```

Comment:
Function source code is not always possible to retrieve and `inspect.getsource` will raise an error. We can live with this limitation for now, but in this case we need a way to bypass caching.

Reply:
do you know how/which cases it will fail to do inspect.getsource? I can add some checker there

Reply:
For now I have it on try and except not the nicest aesthetically but should do the job!

Reply:
The main case is when function constructed dynamically via `eval`/`exec`. It is also possible to distribute python programs as bytecode (`.pyc`) without source code.